### PR TITLE
Integrate codecov and add complete python3.7 support

### LINKED ISSUE #1: Integrate some coverage tool
Integrate a coverage tool like codecov or coveralls. Adapt tox and travis to automatically test with coverage. Provide instructions on integrating this tool w GitHub project.

### LINKED ISSUE #22: Update tox.ini structure
- consider use of skip_sdist key in non-test environments such as doc and lint runs
  - if this is added, comment on this issue before closing with the reason this key is helpful, written for someone who does not know much about tox
- for the default test command, use `/usr/bin/env make test` instead of invoking pytest directly
  - to the makefile target, add `--cov=./`; users who do not like this behavior can manually delete it
  - to the makefile target, consider providing `--basetemp` explicitly and using an option that can be set by an `ENVTMPDIR` environment variable (or similar) to properly interface with tox

### LINKED ISSUE #23: Resort test/dev dependencies
Move coverage and tox to the dev_requirements section (from test_requirements currently), as they are not needed by a user who installs the package from pip and runs `make test` or `python setup.py test`

### LINKED ISSUE #24: support python 3.6/3.7 on travis
For python 3.7, provide travis support using 
```
matrix:
  include:
    - python: 3.7
      dist: xenial
      sudo: required
```

See https://github.com/travis-ci/travis-ci/issues/9069#issuecomment-425720905

As we have experienced deploy errors on the xenial dist required for python 3.7, make sure to deploy on python 3.6 only, i.e.

```
deploy:
  - provider: pages
    on:
      python: 3.6
```

### LINKED ISSUE #32: add `-r` flag to `tox` in `make test-all`
Current `make test-all` target calls `tox`, which might not recreate the testing environment if it already exists.
If dependencies have been changed since the last time tox was ran and a `make clean` has not been used since, this might imply that the new version of the core runs on the old environment without reinstalling it.

To fix this, we can consider adding the `-r` flat to the `tox` command, which will force `tox` to re-create the environment every time it is run.

### PR DESCRIPTION
**Pull request by csala**
Resolve #1: Integrate Codecov. Modified test requirements to add pytest-cov, and .travis.yml and tox.ini to add the necessary config.
Resolve #22 : changed the tox.ini structure to use make commands and extras.
Resolve #23 : added tox and coverage as development dependencies instead of test dependencies.
Resolve #24 : added the necessary configuration to support python3.7 on travis. For this, I had to add both the 3.7 config and the `dist: trusty`, otherwise the base image becomes `xenial` and the gh-pages deployment fails unless the branch exists beforehand.
Resolve #32 : Add `-r` flag to `make test-all` command to have `tox` regenerate the environment everytime it is is run.

### CODE REVIEW

File: {{cookiecutter.repository_name}}/setup.py [hunk 1]
```diff
@@ -11,22 +11,21 @@
...
+setup_requires = [
     'pytest-runner>=2.11.1',
 ]
 
-test_requirements = [
-    'coverage>=4.5.1',
+tests_require = [
     'pytest>=3.4.2',
-    'tox>=2.9.1',
+    'pytest-cov>=2.6.0',
```
Comment (micahjsmith) [inline {{cookiecutter.repository_name}}/setup.py:25]: why is this required under tests_require? should it go in development_requires?

File: cookiecutter.json [hunk 2]
```diff
@@ -15,6 +15,7 @@
   "support_py2": "n",
   "use_pypi_deployment_with_travis": "y",
   "use_ghpages_deployment_with_travis": "y",
+  "integrate_codecov": "y",
```
Comment (micahjsmith) [inline cookiecutter.json]: rename to `"use_codecov"` for consistency with the other similar options

File: {{cookiecutter.repository_name}}/setup.py (hunk 1 above)
Comment (csala) [reply {{cookiecutter.repository_name}}/setup.py:25]: It is needed here, because the `make test` command now includes de `--cov` part, and the `tox` config installs `dev` extras only for docs and lint. The default test setup installs only the `test` requirements.

File: cookiecutter.json (hunk 2 above)
Comment (csala) [reply cookiecutter.json]: Fair point. Changing it!

File: {{cookiecutter.repository_name}}/setup.py (hunk 1 above)
Comment (micahjsmith) [reply {{cookiecutter.repository_name}}/setup.py:25]: ok

**Review by micahjsmith — APPROVED** [review]
